methods.py

```diff
@@ -81,6 +80,34 @@ def add_source_files(self, sources, files, allow_gen=False):
     return True
 
 
+def redirect_emitter(target, source, env):
+    """
+    Emitter to automatically redirect object/library build files to the `bin/obj` directory,
+    retaining subfolder structure. External build files will attempt to retain subfolder
+    structure relative to their environment's parent directory, sorted under `bin/obj/external`.
+    If `redirect_build_objects` is `False`, or an external build file isn't relative to the
+    passed environment, this emitter does nothing.
+    """
+    if not env["redirect_build_objects"]:
+        return target, source
+    redirected_targets = []
+    for item in env.Flatten(target):
+        if isinstance(item, str):
+            item = env.File(item)
+        path = Path(item.get_abspath()).resolve()
+        try:
+            item = env.File(f"#bin/obj/{path.relative_to(base_folder)}")
+        except ValueError:
+            # External file; most likely a custom module. Assume relevant environment was passed.
+            alt_base = Path(env.Dir(".").get_abspath()).parent.resolve()
+            try:
+                item = env.File(f"#bin/obj/external/{path.relative_to(alt_base)}")
+            except ValueError:
+                pass
+        redirected_targets.append(item)
```

Comment:
I'm not sure I understand where the `ValueError` would be thrown. Is it `env.File` constructor, for files pointing outside of `..`?

Let's assume for a second that `path` points somewhere outside base_folder (e.g. absolute path to `../godot-libs/folder`.
Your string may evaluate to `f"#bin/obj/../var/hacky_target"` which would evaluate to `f"#bin/godot-libs/folder"`.

I would encourage sanity checking the input (also to make assumptions clear):
```py
if base_folder in path.parents:
    # ... Place in bin/obj
else:
    # ... Place in bin/obj/external
```

Not sure if that fully works, but I hope it gets across what I mean :)

Reply:
They'd be thrown for `path.relative_to` if it isn't possible to setup a relative path. It definitely reeks of codesmell, so checking `parents` is a **much** cleaner way to achieve this effect; will integrate now!